server/src/main/java/keywhiz/service/daos/ClientDAO.java

```diff
@@ -138,26 +166,69 @@ public void sawClient(Client client, @Nullable Principal principal) {
   }
 
   public Optional<Client> getClientByName(String name) {
-    ClientsRecord r = dslContext.fetchOne(CLIENTS, CLIENTS.NAME.eq(name));
-    return Optional.ofNullable(r).map(clientMapper::map);
+    return getClient(CLIENTS.NAME.eq(name));
   }
 
   public Optional<Client> getClientBySpiffeId(URI spiffeId) {
-    ClientsRecord r = dslContext.fetchOne(CLIENTS, CLIENTS.SPIFFE_ID.eq(spiffeId.toASCIIString()));
-    return Optional.ofNullable(r).map(clientMapper::map);
+    return getClient(CLIENTS.SPIFFE_ID.eq(spiffeId.toASCIIString()));
   }
 
   public Optional<Client> getClientById(long id) {
-    ClientsRecord r = dslContext.fetchOne(CLIENTS, CLIENTS.ID.eq(id));
-    return Optional.ofNullable(r).map(clientMapper::map);
+    return getClient(CLIENTS.ID.eq(id));
+  }
+
+  private Optional<Client> getClient(Condition condition) {
+    Record record = dslContext
+        .select(CLIENTS.fields())
+        .select(CLIENT_OWNERS.ID, CLIENT_OWNERS.NAME)
+        .from(CLIENTS)
+        .leftJoin(CLIENT_OWNERS)
+        .on(CLIENTS.OWNER.eq(CLIENT_OWNERS.ID))
+        .where(condition)
+        .fetchOne();
+
+    return Optional.ofNullable(recordToClient(record));
+  }
+
+  private Client recordToClient(Record record) {
+    if (record == null) {
+      return null;
+    }
+
+    ClientsRecord clientRecord = record.into(CLIENTS);
+    GroupsRecord ownerRecord = record.into(CLIENT_OWNERS);
+
+    boolean danglingOwner = clientRecord.getOwner() != null && ownerRecord.getId() == null;
+    if (danglingOwner) {
+      throw new IllegalStateException(
+          String.format(
+              "Owner %s for client %s is missing.",
+              clientRecord.getOwner(),
+              clientRecord.getName()));
+    }
+
+    Client client = clientMapper.map(clientRecord);
+    if (ownerRecord != null) {
+      client.setOwner(ownerRecord.getName());
+    }
+
+    return client;
   }
 
   public ImmutableSet<Client> getClients() {
-    List<Client> r = dslContext
-        .selectFrom(CLIENTS)
-        .fetch()
-        .map(clientMapper);
-    return ImmutableSet.copyOf(r);
+    Result<Record> results = dslContext
+        .select(CLIENTS.fields())
+        .select(CLIENT_OWNERS.NAME)
+        .from(CLIENTS)
+        .leftJoin(CLIENT_OWNERS)
+        .on(CLIENTS.OWNER.eq(CLIENT_OWNERS.ID))
+        .fetch();
+
+    Set<Client> clients = results.stream()
+        .map(this::recordToClient)
+        .collect(Collectors.toSet());
+
+    return ImmutableSet.copyOf(clients);
```

Comment:
You can map off of results instead if you want.

```suggestion
    List<Client> clients = dslContext
        .select(CLIENTS.fields())
        .select(CLIENT_OWNERS.NAME)
        .from(CLIENTS)
        .leftJoin(CLIENT_OWNERS)
        .on(CLIENTS.OWNER.eq(CLIENT_OWNERS.ID))
        .fetch()
        .map(this::recordToClient);

    return ImmutableSet.copyOf(clients);
```